### PR TITLE
API to find funding utxos in psbt

### PR DESCRIPTION
**Pull request by violet360**
### Current status
The API returns a vector of UTXOs and has return type `Result<Vec<&TxOut>, Error>`

### Expected
The return statement should be of type `sighash::Prevouts` as pointed in #849

### CODE REVIEW

**Comment by Eunoia1729** [comment]
Nit: Next time you may avoid adding/removing blank lines.

**Review by tcharding — COMMENTED** [review]
High level review only, thanks!

File: src/util/psbt/mod.rs [hunk 1]
```diff
@@ -72,6 +72,28 @@ pub struct PartiallySignedTransaction {
...
+                } else if let Some(ref non_witness_utxo) = psbt_input.non_witness_utxo {
+                     // get output index (vout) from the outpoint and use it to get the non_witness_utxo
+                    let vout = self.unsigned_tx.input[i].previous_output.vout as usize;
+                    prev_outs.push(&non_witness_utxo.output[vout])
+                } else {
+                    // created MissingUtxo error enum in errors.rs
+                    return Err(Error::MissingUtxo);
+                }
+            }
+            Ok(prev_outs)
```
[the remark this answers is not in the document]
Comment (Kixunil) [reply src/util/psbt/mod.rs:94]: Definitely *not*, that'd defeat the pointof iterator which is to avoid allocation. The whole function must be one giant iterator - presumably `zip()`ing together inputs and mapping them.

Also I just realized the PSBT type is **not** guaranteed to have inputs of same length. The current implementation panics implicitly in **some** unequal-len situations which isn't documented. I think we should just add `assert_eq!()` at the top and document that the function panics if lengths are not equal. Long term we will change the API to statically guarantee same lengths. I believe @dr-orlovsky is working on that.

[the remark this answers is not in the document]
Comment (tcharding) [reply src/util/psbt/mod.rs:94]: Cool, cheers.

[the remark this answers is not in the document]
Comment (violet360) [reply src/util/psbt/mod.rs:94]: @Kixunil and @tcharding , iterators are pretty cool but returning `impl Iterator<item=R<T, E>>` seems a little tricky, tried to run locally but needed itertools, [see this](https://users.rust-lang.org/t/like-early-return-on-iterator-error/27589/11), we don't want to install extra dependencies, is there a way to skip itertools here ?

[the remark this answers is not in the document]
Comment (violet360) [reply src/util/psbt/mod.rs:94]: `zip()`ped psbt inputs and transaction inputs, [code](https://pastebin.com/FkzTKAa7), this one gives error on line12 (inside `else{}`), then used `itertools::process_results` which claimed to handle `Iterator<Item = Result<T, E>>` and it worked without error, but it required me to install an extra dependency, which I don't think is a good idea.....is there a better way ?
could have used filter_map() and partition() but former just ignored the errors and latter would partition Ok and Err but I want the code to throw error once the first error is detected.

File: src/util/psbt/mod.rs [hunk 2]
```diff
@@ -72,27 +72,22 @@ pub struct PartiallySignedTransaction {
...
-    pub fn to_prevouts(&self) -> Result<Vec<&TxOut>, Error> {
-        let mut prev_outs : Vec<&TxOut> = Vec::new(); 
-        for i in 0..self.inputs.len() {
-            let psbt_input = &self.inputs[i];
-             // check for witness utxos
-                if let Some(ref witness_utxo) = psbt_input.witness_utxo {
-                    prev_outs.push(&witness_utxo)
-                     // if no witness utxos then catch the non witness part of the input
+    
+    /// Returns an iterator for the funding UTXOs of the psbt
```
Comment (violet360) [inline src/util/psbt/mod.rs:76]: documentation can be improved in IMO, but not sure.

File: src/util/psbt/error.rs [hunk 3]
```diff
@@ -37,6 +37,9 @@ pub enum Error {
     /// Magic bytes for a PSBT must be the ASCII for "psbt" serialized in most
     /// significant byte order.
     InvalidMagic,
+    ///
```
Comment (Kixunil) [inline src/util/psbt/error.rs:40]: Blocker: Missing doc comment

File: src/util/psbt/error.rs [hunk 4]
```diff
@@ -98,6 +101,7 @@ impl fmt::Display for Error {
             }
             Error::NoMorePairs => f.write_str("no more key-value pairs for this psbt map"),
             Error::HashParseError(e) => write!(f, "Hash Parse Error: {}", e),
+            Error::MissingUtxo => f.write_str("Utxo is missing"),
```
Comment (Kixunil) [inline src/util/psbt/error.rs:104]: Rewording suggestion: "UTXO information is not present in PSBT"
This conveys a bit more context that may be needed for debugging.

File: src/util/psbt/mod.rs [hunk 5]
```diff
@@ -72,6 +72,24 @@ pub struct PartiallySignedTransaction {
 }
 
 impl PartiallySignedTransaction {
+    /// Returns an iterator for the funding UTXOs of the psbt
+    /// function throws error if transaction inputs have unequal lengths
```
Comment (Kixunil) [inline src/util/psbt/mod.rs:76]: It doesn't "throw" and error though. It panics. In Rust the usual style of documenting panics is like this:

```rust
/// Returns an iterator for the funding UTXOs of the psbt
///
/// For each PSBT input that contains UTXO information `Ok` is returned containing that information.
/// The order of returned items is same as the order of inputs.
///
/// ## Errors
///
/// The function returns error when UTXO information is not present or is invalid.
///
/// ## Panics
///
/// The function panics if the length of transaction inputs is not equal to the length of PSBT inputs.
```

(See below for "or invalid" case.)

File: src/util/psbt/mod.rs [hunk 6]
```diff
@@ -72,6 +72,24 @@ pub struct PartiallySignedTransaction {
...
+    pub fn get_funding_utxos(&self) -> impl Iterator<Item = Result<&TxOut, Error>> {
+        assert_eq!(self.inputs.len(), self.unsigned_tx.input.len());
+        self.unsigned_tx.input.iter().zip(&self.inputs).map(|(tx_input, psbt_input)| {
+            if let Some(ref witness_utxo) = psbt_input.witness_utxo {
```
Comment (Kixunil) [inline src/util/psbt/mod.rs:80]: Note: you can avoid `ref` by borrowing: `if let Some(witness_utxo) = &psbt_input.witness_utxo`
I personally prefer this style and I think it's used in this library but not a big deal. `ref` has it's own appeal too.

Also perhaps using match on the whole thing would be clearer:

```rust
match (&psbt_input.witness_utxo, &psbt_input.non_witness_utxo) {
    (Some(witness_utxo), _) => Ok(witness_utxo),
    (None, Some(non_witness_utxo) => {
        ...
    },
    (None, None) => Err(Error::MissingUtxo),
}
```

File: src/util/psbt/mod.rs [hunk 7]
```diff
@@ -72,6 +72,24 @@ pub struct PartiallySignedTransaction {
...
+                    Ok(witness_utxo)
+                    // if no witness utxos then catch the non witness part of the input
+                } else if let Some(ref non_witness_utxo) = psbt_input.non_witness_utxo {
+                        // get output index (vout) from the outpoint and use it to get the non_witness_utxo
+                    let vout = tx_input.previous_output.vout as usize;
+                    Ok(&non_witness_utxo.output[vout])
```
Comment (Kixunil) [inline src/util/psbt/mod.rs:86]: Blocker:

This code is vulnerable to out-of-bounds access if PSBT comes from untrusted source (e.g. BIP78 would be the case).
There's an easy fix:

```rust
non_witness_utxo.output.get(vout).ok_or(Error::PsbtUtxoOutOfbounds)
```

Obviously this requires adding another error variant.

Why it's OK to panic above but not here: the deserializer guarantees inputs will have equal length so that's generally safe and we even plan to enforce it statically. This is not the case here since the deserializer does **not** guarantee `vout` being in bounds.

**Review by Kixunil — CHANGES_REQUESTED** [review]
Thanks, the code looks almost good but we really have to fix two more issues I found.

There's another thing: should we use a dedicated error type? I tend towards yes but it can be done later as well. I suggest @violet360 if it's too overwhelming to you to add another error type leave it for later, if you feel like adding it you can do it within this PR.

FTR, this PR will most likely **not** go to the Taproot release anyway.

File: src/util/psbt/error.rs (hunk 3 above)
Comment (violet360) [reply src/util/psbt/error.rs:40]: Done

File: src/util/psbt/error.rs (hunk 4 above)
Comment (violet360) [reply src/util/psbt/error.rs:104]: Sorted

File: src/util/psbt/mod.rs (hunk 7 above)
Comment (violet360) [reply src/util/psbt/mod.rs:86]: Wouldn't this cause arms mismatch ? 
at one arm `&txout` type, but in the second arm `Result<&txout, error>` type....

File: src/util/psbt/mod.rs (hunk 6 above)
Comment (violet360) [reply src/util/psbt/mod.rs:80]: > Thanks, the code looks almost good but we really have to fix two more issues I found.
> 
> There's another thing: should we use a dedicated error type? I tend towards yes but it can be done later as well. I suggest @violet360 if it's too overwhelming to you to add another error type leave it for later, if you feel like adding it you can do it within this PR.
> 
> FTR, this PR will most likely **not** go to the Taproot release anyway.

yes I too was thinking about generating a seperate error type for psbts but thought it would be an overkill, will try committing that in this PR itself.....

Comment (Kixunil) [reply src/util/psbt/mod.rs:80]: Note that it still makes sense to also have the variants in `Error` and `From<PsbtUtxoError> for Error`

Comment (violet360) [reply src/util/psbt/mod.rs]: yes,got it, btw, just to be on the same page by a dedicated error type you mean a seperate enum , PsbtUtxoError maybe, in error.rs right ? atleast that's what I think.
just like other error types.

Comment (violet360) [reply src/util/psbt/mod.rs]: It will be cool to have seperate error type for psbt, currently we only have two error types in psbt section....`PsbtUtxoOutOfbounds` and `MissingUtxo` declared in error, and perhaps we can club `pub enum PsbtHash` too inside `psbtUtxoError`

File: src/util/psbt/mod.rs [hunk 8]
```diff
@@ -171,6 +197,7 @@ mod display_from_str {
 #[cfg_attr(docsrs, doc(cfg(feature = "base64")))]
 pub use self::display_from_str::PsbtParseError;
 
+
```
Comment (Kixunil) [inline src/util/psbt/mod.rs:200]: This empty line seems unrelated. It'd be nice to remove it during next push.

File: src/util/psbt/error.rs [hunk 9]
```diff
@@ -37,8 +37,13 @@ pub enum Error {
     /// Magic bytes for a PSBT must be the ASCII for "psbt" serialized in most
     /// significant byte order.
     InvalidMagic,
+    ///Missing both the witness and non-witness utxo
+    MissingUtxo,
     /// The separator for a PSBT must be `0xff`.
     InvalidSeparator,
+    /// Triggered when output index goes out of bounds of when trying to get
```
Comment (Kixunil) [inline src/util/psbt/error.rs:44]: This doesn't seem grammatically correct but as a non-native I'll defer this to @tcharding

**Review by Kixunil — COMMENTED** [review]
The code is acceptable now. Just some minor details.

Because it's a new non-essential feature this should be merged after 0.28 release.

File: src/util/psbt/mod.rs (hunk 6 above)
Comment (violet360) [reply src/util/psbt/mod.rs]: should I address it here, or club it up with #837 ?

File: src/util/psbt/error.rs [hunk 10]
```diff
@@ -98,6 +102,8 @@ impl fmt::Display for Error {
             }
             Error::NoMorePairs => f.write_str("no more key-value pairs for this psbt map"),
             Error::HashParseError(e) => write!(f, "Hash Parse Error: {}", e),
+            Error::MissingUtxo => f.write_str("UTXO information is not present in PSBT"),
+            Error::PsbtUtxoOutOfbounds => f.write_str("Output index is out of bounds of non witness script output array"),
```
Comment (Kixunil) [inline src/util/psbt/error.rs:106]: Oh, forgot to comment before that `"Output index` should probably be `"output index` to be consistent with other messages.

**Review by Kixunil — COMMENTED** [review]
Looks great, you just need to squash the commits into one.

File: src/util/psbt/error.rs [hunk 11]
```diff
@@ -37,8 +37,12 @@ pub enum Error {
     /// Magic bytes for a PSBT must be the ASCII for "psbt" serialized in most
     /// significant byte order.
     InvalidMagic,
+    ///Missing both the witness and non-witness utxo
```
Comment (tcharding) [inline src/util/psbt/error.rs:40]:
```suggestion
    /// Missing both the witness and non-witness utxo.
```

File: src/util/psbt/error.rs [hunk 12]
```diff
@@ -37,8 +37,12 @@ pub enum Error {
...
+    MissingUtxo,
     /// The separator for a PSBT must be `0xff`.
     InvalidSeparator,
+    /// Triggered when output index goes out of bounds of the output in non witness utxo
```
Comment (tcharding) [inline src/util/psbt/error.rs:44]:
```suggestion
    /// Triggered when output index does not exist in PSBT input (non-witness) UTXO.
```

Not sure if the content of this comment is correct but it shows a grammatical improvement, please modify as you see fit.

File: src/util/psbt/error.rs (hunk 10 above)
Comment (tcharding) [reply src/util/psbt/error.rs:106]: `Hash Parse Error:` is wrong too, a patch put at the front of this PR cleaning that all up would be super useful if you can be bothered. Don't feel obliged though. You will likely touch this line when the error comment is changed anyways.

File: src/util/psbt/mod.rs [hunk 13]
```diff
@@ -20,14 +20,14 @@
...
-use blockdata::transaction::Transaction;
+use blockdata::transaction::{ TxOut, Transaction};
 use consensus::{encode, Encodable, Decodable};
 use consensus::encode::MAX_VEC_SIZE;
+pub use util::sighash::Prevouts;
 
 use prelude::*;
 
 use io;
-
```
Comment (tcharding) [inline src/util/psbt/mod.rs:30]: Please no 'random' whitespace changes. The reason is that it makes reviewing the PR harder because reviewers must stop to think what is this, why is it being done, is it correct, is it a step backwards in readability. All easy questions but unnecessary for getting your work merged. The easier you make it for reviewers to ack your work the easier you'll get PR's merged, that's a win for everyone :)

**Review by tcharding — COMMENTED** [review]
Just a few comments, thanks.

File: src/util/psbt/error.rs (hunk 12 above)
Comment (Kixunil) [reply src/util/psbt/error.rs:44]: While strictly speaking correct, thinking of it in terms of "out of bounds" feels more natural to me. Won't block if others prefer your suggestion.

Comment (tcharding) [reply src/util/psbt/error.rs:44]: Fair comment, this is a tricky little rustdoc to write succinctly.

Comment (violet360) [reply src/util/psbt/error.rs:44]: out of bound seemed natural to me too, but whatever you guys prefer at the end though.

File: src/util/psbt/mod.rs (hunk 13 above)
Comment (violet360) [reply src/util/psbt/mod.rs:30]: Yeah, silly mistake of mine, do you guys use any sort of linter ?

Comment (violet360) [reply src/util/psbt/mod.rs:30]: Yeah, silly mistake of mine ... will take care of it from next commits.

Comment (tcharding) [reply src/util/psbt/mod.rs:30]: Cheers mate, no stress. If I had a dollar for ever mistake I made in this repo in the last three months I'd be fiat rich :)

File: src/util/psbt/error.rs (hunk 12 above)
Comment (tcharding) [reply src/util/psbt/error.rs:44]: How's this?
```
    /// Triggered when output index is out of bounds in relation to the output in non-witness UTXO.
```

Comment (Kixunil) [reply src/util/psbt/error.rs:44]: Sounds good. But should it be "triggered"? Maybe "returned is better"?

Comment (violet360) [reply src/util/psbt/error.rs:44]: I think in case of error words like triggered, panic, etc. Sounds more suitable than "returned", don't you think ?

Comment (Kixunil) [reply src/util/psbt/error.rs:44]: My mental model in Rust is that errors are returned (well, they actually are :)). But honestly, I don't care much if others find "triggered" better.

Comment (tcharding) [reply src/util/psbt/error.rs:44]: FWIW we have 'triggered' in one other place in the repo
```
git grep trig                                                                                                                                                              
src/blockdata/script.rs:1279:        // which triggerred overflow bugs on 32-bit machines in script formatting in the past.
```

I'm not fussed either way.

File: src/util/psbt/mod.rs [hunk 14]
```diff
@@ -72,6 +72,32 @@ pub struct PartiallySignedTransaction {
...
+    ///
+    /// ## Errors
+    ///
+    /// The function returns error when UTXO information is not present or is invalid.
+    ///
+    /// ## Panics
+    ///
+    /// The function panics if the length of transaction inputs is not equal to the length of PSBT inputs.
+    pub fn get_funding_utxos(&self) -> impl Iterator<Item = Result<&TxOut, Error>> {
+        assert_eq!(self.inputs.len(), self.unsigned_tx.input.len());
```
Comment (dr-orlovsky) [inline src/util/psbt/mod.rs:88]:
```suggestion
        assert_debug_eq!(self.inputs.len(), self.unsigned_tx.input.len());
```

**Review by dr-orlovsky — COMMENTED** [review]
Agree with other comments, just adding my 5 cents

File: src/util/psbt/mod.rs (hunk 14 above)
Comment (violet360) [reply src/util/psbt/mod.rs]: Hi !
you meant `debug_assert_eq!` right ?

Comment (dr-orlovsky) [reply src/util/psbt/mod.rs]: Yes, sorry.

File: src/util/psbt/mod.rs [hunk 15]
```diff
@@ -72,6 +72,32 @@ pub struct PartiallySignedTransaction {
...
+    /// The order of returned items is same as the order of inputs.
+    ///
+    /// ## Errors
+    ///
+    /// The function returns error when UTXO information is not present or is invalid.
+    ///
+    /// ## Panics
+    ///
+    /// The function panics if the length of transaction inputs is not equal to the length of PSBT inputs.
+    pub fn get_funding_utxos(&self) -> impl Iterator<Item = Result<&TxOut, Error>> {
```
Comment (sanket1729) [inline src/util/psbt/mod.rs:87]: On a similar theme for #861, I think we rename this to `iter_funding_utxos`?

**Review by sanket1729 — COMMENTED** [review]
Sorry for the delayed review. This is a great shape after several iterations and thanks for addressing all the feedback @violet360 . 

Everything is good from my side, just the final API naming.

File: src/util/psbt/mod.rs (hunk 15 above)
Comment (violet360) [reply src/util/psbt/mod.rs:87]: yes, consider it done 👍🏻

**Comment by Kixunil** [comment]
@dr-orlovsky I disagree with it being debug assert. If the two have different length it's a clear inconsistency of input data and there's no good solution. Debug assert will just allow silent weird bugs in release.

Regarding naming, `iter_` is acceptable but unnecessary. `str::bytes()` nor `str::chars()` have `iter_` prefix and it's fine. Agree with dropping `get_` prefix. `funding_utxos` looks idiomatic and reads well to me.

**Comment by violet360** [comment]
@Kixunil and @sanket1729  I also want to improve the error management of psbt, I think there is a lot of room for that, planning to get that done in this PR itself, or maybe a new PR.
1. Should I create a separate enum `psbt` and add `PsbtHash` as the child enum, and add `psbt` as a child enum of `Error`, kinda like :
```rust
pub enum psbt {
...
  PsbtHash, // psbthash child enum of psbt
/// Missing both the witness and non-witness utxo.
  MissingUtxo,
/// Returned when output index is out of bounds in relation to the output in non-witness UTXO.
  PsbtUtxoOutOfbounds,
..
}
...
pub enum Error {
...
  psbt, 
...
}
```
2. seems like `use psbt::raw` <--> `psbt::Error` and `consensus::encode`<--> `psbt::Error` have a circular dependencies, do we need to address that?

**Review by sanket1729 — DISMISSED** [review]
ACK 9f523e0747d551eeeecb634b425701ceb965c497. If there are further iterations, feel free to request review from me. 
> @dr-orlovsky I disagree with it being debug assert. If the two have different length it's a clear inconsistency of input data and there's no good solution. Debug assert will just allow silent weird bugs in release.

Indeed, asserts should be `assert!` instead of `debug_assert!` unless it benchmarked that it affects performance. I had a similar issue in taproot review. 
https://github.com/rust-bitcoin/rust-bitcoin/pull/677#discussion_r744290472

Perhaps, that should also be updated to `assert!` and we can inspect the codebase for misc `debug_assert!` and change those to `assert!`

**Comment by sanket1729** [comment]
@violet360, we should take this discussion to #837

**Review by dr-orlovsky — DISMISSED** [review]
utACK 9f523e0747d551eeeecb634b425701ceb965c497

**Comment by dr-orlovsky** [comment]
@sanket1729 sorry, IDK why but your comment does not displays to me.

@Kixunil 
My understanding is that whenever there is _internal error in the logic of the code_, it should be `debug_assert*`. `assert*`s are for tests, and if there might be a error in the input data, than it should be a `Result` with error type returned.

Here, if the supplied PSBT (as an input) have different number of inputs in tx and map, it must error at the moment it was unparsed into `PartiallySignedTransaction` structure. And if it was not, it is an internal program error.

**Comment by sanket1729** [comment]
@dr-orlovsky See: https://doc.rust-lang.org/std/macro.debug_assert.html

Quote from the docs: 
> Replacing [assert!](https://doc.rust-lang.org/std/macro.assert.html) with debug_assert! is thus only encouraged after thorough profiling, and more importantly, only in safe code!

**Comment by dr-orlovsky** [comment]
@Kixunil seems like all your comments were addressed. Can you confirm so we can merge this one?

**Comment by Kixunil** [comment]
I view the logic bug similar to indexing out of bounds or calling `copy_from_slice` on non-equal-length slices. These always panic. If the type guaranteed consistency then I'd be in favor of `debug_assert!`. The code currently can not guarantee consistency because the fields are public. You can construct inconsistent PSBT by just forgetting to push stuff. We could also return `Err` instead of panicking but this kind of error can only be caused by buggy code, not input data.

You could argue this is not memory-safety issue but I view bugs in bitcoin-related code as seriously as memory safety. People could lose sats.

V2 API will remove this by construction but until then `assert_eq!` should be the most idiomatic approach.

**Comment by violet360** [comment]
sorry for the late comment...was busy with some other stuff lately
it seems there is a disagreement at the panic statement, altho I too lean towards `!assert_eq` but not by experience but by the docs as pointed out by @sanket1729 .
can we conclude?

**Comment by tcharding** [comment]
> it seems there is a disagreement at the panic statement

FWIW, from my reading of this thread there is consensus on `assert_eq!`.

**Review by Kixunil — APPROVED** [review]
ACK 5afb0eaf40642bfda87f2681a650a919d43163a4

**Review by sanket1729 — APPROVED** [review]
ACK 5afb0eaf40642bfda87f2681a650a919d43163a4. Thanks for being patient with this.

**Comment by DanGould** [comment]
This interface does not return complete information about the funding utxo. it only returns psbt v0 input map info: `TxOut` which has script_pubkey and value. Unfortunately, txid, vout, sequence number, and locktime are still only the global map's unsigned_tx. So if for example you want to check some psbt set of input utxos against those in your wallet, as is a standard receiver payjoin check, you need data beyond what this function returns.

It would be nice in the PSBTv2 implementation to return an `Input` map element with that complete set of data rather than just a `TxOut { script_pubkey, locktime}`
